src/palimpzest/utils/progress.py

```diff
@@ -47,179 +40,230 @@ def get_memory_usage() -> float:
+            #TextColumn("[cyan]Mem: {task.fields[memory]:.1f}MB"),
+            TextColumn("[green]Cost: ${task.fields[cost]:.4f}"),
             TextColumn("\n[white]{task.fields[recent]}"),  # Recent text on new line
             refresh_per_second=10,
             expand=True,   # Use full width
         )
-        self.task_id = None
-        
-    def start(self, total: int):
-        # Print a newline before starting to separate from previous output
-        print()
-        
-        self.task_id = self.progress.add_task(
-            "Processing", 
+
+        # initialize mapping from op_id --> ProgressStats
+        self.op_id_to_stats: dict[str, ProgressStats] = {}
+
+        # initialize mapping from op_id --> task
+        self.op_id_to_task = {}
+
+        # initialize start time
+        self.start_time = None
+
+        # create mapping from op_id --> next_op
+        self.op_id_to_next_op: dict[str, PhysicalOperator] = {}
+        for op_idx, op in enumerate(plan.operators):
+            op_id = op.get_op_id()
+            next_op = plan.operators[op_idx + 1] if op_idx + 1 < len(plan.operators) else None
+            self.op_id_to_next_op[op_id] = next_op
+
+        # compute the total number of inputs to be processed by the plan
+        datareader_len = len(plan.operators[0].datareader)
+        total = datareader_len if num_samples is None else min(num_samples, datareader_len)
+
+        # add a task to the progress manager for each operator in the plan
+        for op in plan.operators:
+            # get the op id and a short string representation of the op; (str(op) is too long)
+            op_id = op.get_op_id()
+            op_str = f"{op.op_name()} ({op_id})"
+
+            # update the `total` if we encounter an AggregateOp or LimitScanOp
+            if isinstance(op, AggregateOp):
+                total = 1
+            elif isinstance(op, LimitScanOp):
+                total = op.limit
+
+            self.add_task(op_id, op_str, total)
+
+    def get_task_total(self, op_id: str) -> int:
+        """Return the current total value for the given task."""
+        task = self.op_id_to_task[op_id]
+        return self.progress._tasks[task].total
+
+    def get_task_description(self, op_id: str) -> str:
+        """Return the current description for the given task."""
+        task = self.op_id_to_task[op_id]
+        return self.progress._tasks[task].description
+
+    @abstractmethod
+    def add_task(self, op_id: str, op_str: str, total: int):
+        """Initialize progress tracking for operator execution with total items"""
+        pass
+
+    @abstractmethod
+    def start(self):
+        """Start the progress bar(s)"""
+        pass
+
+    @abstractmethod
+    def incr(self, op_id: str, num_outputs: int = 1, display_text: str | None = None, **kwargs):
+        """
+        Advance the progress bar for the given operator by one. Modify the downstream operators'
```

Comment:
Advance the progress bar for the given operator by num_outputs?

Reply:
Not exactly; the semantics of this function are that every time it is executed we advance the progress bar by 1. This is because the progress bar represents what fraction of the **inputs** have been processed by that operator.

`num_outputs` specifies how many outputs were generated by the operator when processing the input for which `incr()` was called. E.g. a filter which filters an input record will advance its progress bar by 1, but the next operator will now have 1 fewer inputs to process. Alternatively, a convert which generates 3 `num_outputs` will increase the inputs for the next operator by `delta = num_outputs - 1 = 2`